lang/src/main/kotlin/org/partiql/lang/util/NumberExtensions.kt

```diff
@@ -256,3 +257,302 @@ val Number.isPosInf get() = when (this) {
+}
+
+/**
+ * Computes the nth root of a given BigDecimal x.
+ * where x needs to be positive.
+ */
+private fun BigDecimal.intRoot(
+    index: Int,
+    mathContext: MathContext
+): BigDecimal {
+    val operationMC = MathContext(
+        if (mathContext.precision + 2 < 0) Int.MAX_VALUE else mathContext.precision + 2,
+        mathContext.roundingMode
+    )
+
+    val tolerance: BigDecimal = BigDecimal.valueOf(5L).movePointLeft(mathContext.precision)
+
+    // using Newton's method
+    // x_i = ( (r-1) (x_i-1) ^ r + n) / r (x_i-1) ^(r-1)
+    // where n is the number whose rth root we want to compute
+    val r = BigDecimal.valueOf(index.toLong())
+    val rMinusOne = BigDecimal.valueOf(index.toLong() - 1L)
+
+    // The initial approximation is x/index.
+    var res = this.divide(r, mathContext)
+
+    var resPrev: BigDecimal
+    do {
+        // x^(index-1)
+        val xToRMinusOne = res.pow(index - 1, operationMC)
+        // x^index
+        val xToR = res.multiply(xToRMinusOne, operationMC)
+
+        // n + (index-1)*(x^index)
+        val numerator = this.add(rMinusOne.multiply(xToR, operationMC), operationMC)
+
+        // (index*(x^(index-1))
+        val denominator = r.multiply(xToRMinusOne, operationMC)
+
+        // x = (n + (index-1)*(x^index)) / (index*(x^(index-1)))
+        resPrev = res
+
+        res = numerator.divide(denominator, operationMC)
+    } while (res.round(mathContext).subtract(resPrev.round(mathContext)).abs() > tolerance)
+    return res
+}
+
+/**
+ * Computes the square root of a given BigDecimal.
+ * See https://dl.acm.org/doi/pdf/10.1145/214408.214413
+ */
+fun BigDecimal.squareRoot(mathContext: MathContext = MATH_CONTEXT): BigDecimal {
+    if (this.signum() < 0) {
+        throw ArithmeticException("Cannot take root of a negative number")
+    }
+
+    // We want to utilize the floating number's sqrt method to take an educated guess
+    // to make sure the number is representable
+    // we operate on normalized mantissa, which is [0.1, 10)
+    val stripped = this.stripTrailingZeros()
+    val scale = stripped.scale() - stripped.precision() + 1
+    val scaleAdj = if (scale % 2 == 0) scale else scale - 1
+    val mantissa = stripped.scaleByPowerOfTen(scaleAdj)
+
+    val guess = BigDecimal.valueOf(kotlin.math.sqrt(mantissa.toDouble()))
+
+    // Conservative guess of the precision of the result of a floating point calculation.
+    var guessPrecision = 10
+
+    // we need this additional logic in case of overflow
+    val targetPrecision = mathContext.precision
+    val normalizedPrecision = mantissa.precision()
+    var approx = guess
+
+    val zeroPointFive = BigDecimal.ONE.divide(BigDecimal.valueOf(2))
+    do {
+        // plus 2 for precision buffering
+        val operatingPrecision = kotlin.math.max(
+            kotlin.math.max(guessPrecision, targetPrecision + 2),
+            normalizedPrecision
+        )
+        val tempMC = MathContext(operatingPrecision, RoundingMode.HALF_EVEN)
```

Comment:
Do we have to create a MathContext every time? This creation is likely to slow things down. I see how the precision is changing though.

Reply:
In this specific case, I believe the benefit outweighed the cost. Because bigDecimal operation's run time increase non-linearly w.r.t precision.